qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -912,23 +913,220 @@ def __call__(self, target, basis_fidelity=None, *, _num_basis_uses=None) -> Quan
+                return_circuit.append(self.gate, [q[0], q[1]])
+            return_circuit.compose(decomposition_euler[2 * best_nbasis], [q[0]], inplace=True)
+            return_circuit.compose(decomposition_euler[2 * best_nbasis + 1], [q[1]], inplace=True)
         return return_circuit
 
+    def _get_sx_vz_2cx_efficient_euler(self, decomposition, target_decomposed):
+        """
+        Decomposition of SU(4) gate for device with SX, virtual RZ, and CNOT gates assuming
+        two CNOT gates are needed.
+        """
+        best_nbasis = 2  # by assumption
+        from scipy.spatial.transform import Rotation
+
+        num_1q_uni = len(decomposition)
+        dq0 = np.empty((int(num_1q_uni / 2), 3), dtype=float)
+        dq1 = np.empty((int(num_1q_uni / 2), 3), dtype=float)
+        global_phase = 0.0
+        for iqubit, idecomp in enumerate(range(0, num_1q_uni, 2)):
+            axis, angle, phase = _su2_axis_angle(decomposition[idecomp])
+            rot = Rotation.from_rotvec(axis * angle)
+            euler_angles = rot.as_euler("zxz")
+            dq0[iqubit, :] = euler_angles
+            global_phase += phase
+        for iqubit, idecomp in enumerate(range(1, num_1q_uni, 2)):
+            axis, angle, phase = _su2_axis_angle(decomposition[idecomp])
+            rot = Rotation.from_rotvec(axis * angle)
+            euler_angles = rot.as_euler("xzx")
+            dq1[iqubit, :] = euler_angles
+            global_phase += phase
+        qc = QuantumCircuit(2)
+        qc.global_phase = target_decomposed.global_phase
+        qc.global_phase -= best_nbasis * self.basis.global_phase
+
+        # TODO: make this more effecient
+        qc1q = QuantumCircuit(1)
+        qc1q.rz(dq0[0][0], 0)
+        qc1q.rx(dq0[0][1], 0)
+        qc1q.rz(dq0[0][2] + dq0[1][0], 0)
+        qceuler = self._decomposer1q(Operator(qc1q).data)
+        qceuler.data[-1][0].params[0] += math.pi / 2
+        qc.compose(qceuler, [0], inplace=True)
+        qc1q = QuantumCircuit(1)
+        qc1q.rx(dq1[0][0], 0)
+        qc1q.rz(dq1[0][1], 0)
+        qc1q.rx(dq1[0][2] + dq1[1][0], 0)
+        qceuler = self._decomposer1q(Operator(qc1q).data)
+        qc.compose(qceuler, [1], inplace=True)
+
+        qc.cx(0, 1)
+
+        qc.sx(0)
+        qc.rz(dq0[1][1] - math.pi, 0)
+        qc.sx(0)
+        qc.rz(dq1[1][1], 1)
+
+        qc.cx(0, 1)
+
+        qc1q = QuantumCircuit(1)
+        qc1q.rz(dq0[1][2] + dq0[2][0], 0)
+        qc1q.rx(dq0[2][1], 0)
+        qc1q.rz(dq0[2][2], 0)
+        qceuler = self._decomposer1q(Operator(qc1q).data)
+        qceuler.data[0][0].params[0] += math.pi / 2
+        qc.compose(qceuler, [0], inplace=True)
+        qc1q = QuantumCircuit(1)
+        qc1q.rx(dq1[1][2] + dq1[2][0], 0)
+        qc1q.rz(dq1[2][1], 0)
+        qc1q.rx(dq1[2][2], 0)
+        qceuler = self._decomposer1q(Operator(qc1q).data)
+        qc.compose(qceuler, [1], inplace=True)
+
+        # FIXME: sometimes phase is off by multiple of pi/2
```

Comment:
This is definitely worth tracking down; it surely has a discernible source.

Reply:
agreed